src/FSharpVSPowerTools.Logic/PrintfSpecifiersUsageTagger.fs

```diff
@@ -0,0 +1,115 @@
+﻿namespace FSharpVSPowerTools.PrintfSpecifiersHighlightUsage
+
+open System
+open Microsoft.VisualStudio.Text
+open Microsoft.VisualStudio.Text.Editor
+open Microsoft.VisualStudio.Text.Tagging
+open Microsoft.VisualStudio.Shell.Interop
+open FSharpVSPowerTools
+open FSharpVSPowerTools.ProjectSystem
+open FSharpVSPowerTools.PrintfSpecifiersUsageGetter
+open Microsoft.FSharp.Compiler
+
+type PrintfSpecifiersUsageTag() = 
+    inherit TextMarkerTag(Constants.fsharpPrintfTagType) 
+
+// Reference at http://msdn.microsoft.com/en-us/library/vstudio/dd885121.aspx
+
+/// This tagger will provide tags for every printf format specifier under the cursor.
+type PrintfSpecifiersUsageTagger
+    (
+        textDocument: ITextDocument,
+        view: ITextView, 
+        vsLanguageService: VSLanguageService, 
+        serviceProvider: IServiceProvider,
+        projectFactory: ProjectFactory
+    ) as self =
+    
+    let tagsChanged = Event<_, _>()
+    let mutable spans: NormalizedSnapshotSpanCollection option = None
+    let mutable usages: PrintfSpecifierUse[] option = None
+    let buffer = view.TextBuffer
+
+    let printfSpecifierUseToSpans (u: PrintfSpecifierUse): SnapshotSpan[] =
+        [| fromFSharpRange buffer.CurrentSnapshot u.SpecifierRange
+           fromFSharpRange buffer.CurrentSnapshot u.ArgumentRange |]
+        |> Array.choose id
+    
+    let findUsagesAtPoint (point: SnapshotPoint) (usages: PrintfSpecifierUse[]): PrintfSpecifierUse option =
+        let pos = 
+            Range.mkPos ((point.Snapshot.GetLineNumberFromPosition point.Position) + 1) 
+                        (point.Position - point.GetContainingLine().Start.Position)
+        usages 
+        |> Array.filter (fun x -> 
+            Range.rangeContainsPos x.SpecifierRange pos 
+            || Range.rangeContainsPos x.ArgumentRange pos)
+        |> Array.tryHead
+
+    let onCaretMove (CallInUIContext callInUIContext) =
+        asyncMaybe {
+            let! point = buffer.GetSnapshotPoint view.Caret.Position
+            let! usages = usages
+            let! usages = findUsagesAtPoint point usages
+            let spans = printfSpecifierUseToSpans usages
+            return NormalizedSnapshotSpanCollection spans
+        } 
+        |> Async.bind (fun x ->
+            async {
+                spans <- x
+                return! callInUIContext <| fun _ -> tagsChanged.Trigger(self, SnapshotSpanEventArgs buffer.CurrentSnapshot.FullSpan)
+            }
+        )
+
+    let onCaretMoveListener: DocumentEventListener option ref = ref None
```

Comment:
Can we use `DocumentEventListener lazy` type instead? In the end, onCaretMoveListener is assigned at most once.

```fsharp
onCaretMoveListener := 
                               lazy(new DocumentEventListener ([ViewChange.layoutEvent view; ViewChange.caretEvent view], 200us, onCaretMove))
```